File: Two Pointers/validPalindrome.py
# Given a string s, return true if it is a palindrome, otherwise return false.
# A palindrome is a word that reads the same backward as forward. For example, "racecar" is a palindrome.
# leetcode.com/problems/valid-palindrome

from typing import List

class Solution:
	def isPalindrome(self, s: str) -> bool:
		s = ''.join(e for e in s if e.isalnum()).lower()
		return s == s[::-1]

def main():
	s = "Was it a car or a cat I saw?"
	solution = Solution()
	result = solution.isPalindrome(s)
	print(result)

if __name__ == '__main__':
	main()

File: Two Pointers/validPalindrome.go
// Given a string s, return true if it is a palindrome, otherwise return false.

package main

import "fmt"

type Solution struct{}

func (s *Solution) isPalindrome(str string) bool {
	l, r := 0, len(str)-1
	for l < r {
		if !isAlphanumeric(str[l]) {
			l++
		} else if !isAlphanumeric(str[r]) {
			r--
		} else if toLower(str[l]) != toLower(str[r]) {
			return false
		} else {
			l++
			r--
		}
	}
	return true
}

func isAlphanumeric(ch byte) bool {
	return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || (ch >= '0' && ch <= '9')
}

func toLower(ch byte) byte {
	if ch >= 'A' && ch <= 'Z' {
		return ch + 32
	}
	return ch
}

func main() {
	solution := &Solution{}
	s1 := "Was it a car or a cat I saw?"
	fmt.Printf("Test case 1: %v\n", solution.isPalindrome(s1))

	s2 := "tab a cat"
	fmt.Printf("Test case 2: %v\n", solution.isPalindrome(s2))
}
